src/safeposix/vmmap.rs

```diff
@@ -0,0 +1,626 @@
+        self.entries.iter_mut()
+    }
+
+    // Method to iterate over pages, starting from a given page number
+    fn find_page_iter(
+        &self,
+        page_num: u32,
+    ) -> impl DoubleEndedIterator<Item = (&Interval<u32>, &VmmapEntry)> {
+        if let Some(last_entry) = self.last_entry() {
+            self.entries.overlapping(ie(page_num, last_entry.0.end()))
+        } else {
+            // Return an empty iterator if no last_entry
+            self.entries.overlapping(ie(page_num, page_num))
+        }
+    }
+
+    // Method to iterate over mutable pages, starting from a given page number
+    fn find_page_iter_mut(
+        &mut self,
+        page_num: u32,
+    ) -> impl DoubleEndedIterator<Item = (&Interval<u32>, &mut VmmapEntry)> {
+        if let Some(last_entry) = self.last_entry() {
+            self.entries
+                .overlapping_mut(ie(page_num, last_entry.0.end()))
+        } else {
+            // Return an empty iterator if no last_entry
+            self.entries.overlapping_mut(ie(page_num, page_num))
+        }
+    }
+
+    // Method to find space in the memory map
+    fn find_space(&self, npages: u32) -> Option<Interval<u32>> {
+        let start = self.first_entry();
+        let end = self.last_entry();
+
+        if start == None || end == None {
+            return None;
+        } else {
+            let start_unwrapped = start.unwrap().0.start();
+            let end_unwrapped = end.unwrap().0.end();
+
+            let desired_space = npages + 1; // TODO: check if this is correct
+
+            for gap in self
+                .entries
+                .gaps_trimmed(ie(start_unwrapped, end_unwrapped))
+            {
+                if gap.end() - gap.start() >= desired_space {
+                    return Some(gap);
+                }
+            }
+        }
+
+        None
+    }
+
+    // Method to find space above a hint
+    fn find_space_above_hint(&self, npages: u32, hint: u32) -> Option<Interval<u32>> {
+        let start = hint;
+        let end = self.last_entry();
+
+        if end == None {
+            return None;
+        } else {
+            let end_unwrapped = end.unwrap().0.end();
+
+            let desired_space = npages + 1; // TODO: check if this is correct
+
+            for gap in self.entries.gaps_trimmed(ie(start, end_unwrapped)) {
+                if gap.end() - gap.start() >= desired_space {
+                    return Some(gap);
+                }
+            }
+        }
+
+        None
+    }
+
+    // Method to find map space, rounding page numbers up to the nearest multiple of pages_per_map
+    fn find_map_space(&self, num_pages: u32, pages_per_map: u32) -> Option<Interval<u32>> {
+        let start = self.first_entry();
+        let end = self.last_entry();
+
+        if start == None || end == None {
+            return None;
+        } else {
+            let start_unwrapped = start.unwrap().0.start();
```

Comment:
need to walk through the logic here, should be from end of entry to start of next, but also need to configure 0 and 4GB as endpoints